src/reformat.c

```diff
@@ -129,30 +132,35 @@ avifBool avifPrepareReformatState(const avifImage * image, const avifRGBImage *
     return AVIF_TRUE;
 }
 
+// Formulas 20-31 from https://www.itu.int/rec/T-REC-H.273-201612-I/en
 static int avifReformatStateYToUNorm(avifReformatState * state, float v)
 {
-    int unorm = (int)avifRoundf(v * state->yuvMaxChannelF);
+    int unorm = (int)avifRoundf(v * state->yRange + state->yBias);
     unorm = AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
-    if (state->yuvRange == AVIF_RANGE_LIMITED) {
-        unorm = avifFullToLimitedY(state->yuvDepth, unorm);
-    }
     return unorm;
 }
 
 static int avifReformatStateUVToUNorm(avifReformatState * state, float v)
 {
-    int unorm = (int)avifRoundf(v * state->yuvMaxChannelF);
-    if (state->mode != AVIF_REFORMAT_MODE_IDENTITY) { // Don't adjust U or V when data is raw RGB
-        unorm += state->uvBias;
+    int unorm = 0;
+
+    // For YCgCo, this is only correct for full range
+    if (state->mode == AVIF_REFORMAT_MODE_YCGCO) {
+        assert(state->yuvRange == AVIF_RANGE_FULL);
     }
-    unorm = AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
-    if (state->yuvRange == AVIF_RANGE_LIMITED) {
-        if (state->mode == AVIF_REFORMAT_MODE_IDENTITY) { // use Y range for all channels when data is raw RGB
-            unorm = avifFullToLimitedY(state->yuvDepth, unorm);
-        } else {
-            unorm = avifFullToLimitedUV(state->yuvDepth, unorm);
-        }
+
+    switch (state->mode) {
+        case AVIF_REFORMAT_MODE_YCGCO:
```

Comment:
AVIF_REFORMAT_MODE_YCGCO and AVIF_REFORMAT_MODE_IDENTITY should be handled in the same way. Move this after line 158.

```
        case AVIF_REFORMAT_MODE_IDENTITY:
        case AVIF_REFORMAT_MODE_YCGCO:
            unorm = (int)avifRoundf(v * state->yRange + state->yBias);
            unorm = AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
            break;
```

Reply:
Do you know why the avifyuv test still passes with the AVIF_REFORMAT_MODE_YCGCO case in the wrong place? I wonder if I misunderstood something. Thanks!

Reply:
AVIF_REFORMAT_MODE_YCGCO is not following H.273's procedure, it's just coincidentally we can reuse AVIF_REFORMAT_MODE_YUV_COEFFICIENTS's route to produce mostly same result. Note that current implementation of `avifReformatStateUVToUNorm` also treats AVIF_REFORMAT_MODE_YCGCO the same as AVIF_REFORMAT_MODE_YUV_COEFFICIENTS.

Suppose range=AVIF_RANGE_FULL and mc=AVIF_MATRIX_COEFFICIENTS_YCGCO and depth=8, if strictly follows H.273 we should use Equations 26-28 and 44-46, that is:
| Representation / Operation |  Range |
| --- | --- |
| {E′R,E′G,E′B} | [0., 1.] |
| * 255 | |
| (Not denoted as variable in H.273) | [0., 255.] |
| Clip1Y | |
| {R, G, B} | [0., 255.] |
| Kr-Kb like transformation with Kr = 0.25, Kb = 0.25 | |
| {Y'1, Cg'1, Co'1} (Not denoted as variable in H.273) | Y'1: [0., 255.], Cg'1: [-127.5, 127.5], Co'1: [-127.5, 127.5] |
| Round (rounding halfway cases away from zero) | |
| {Y'2, Cg'2, Co'2} (Not denoted as variable in H.273) | Y'1: [0, 255], Cg'1: [-128, 128], Co'1: [-128, 128] |
| Add bias 128 to Cg'2 and Co'2 | |
| {Y, Cb, Cr} | Y: [0, 255], Cb: [0, 256], Cr: [0, 256] |

libavif's current implementation uses the following procedure:
| Representation / Operation |  Range |
| --- | --- |
| {E′R,E′G,E′B} | [0., 1.] |
| Kr-Kb like transformation with Kr = 0.25, Kb = 0.25 | |
| {Y'3, Cg'3, Co'3} | Y'3: [0., 1.], Cg'3: [-0.5, 0.5], Co'3: [-0.5, 0.5] |
| * 255 (operations from here done in `avifReformatStateUVToUNorm`) | |
| {Y'1, Cb'1, Cr'1} | Y'1: [0., 255.], Cb'1: [-127.5, 127.5], Cr'1: [-127.5, 127.5] |
| Round (rounding halfway cases up, see `avifRoundf`) | |
| {Y'2, Cb'2, Cr'2} | Y'2: [0, 255], Cb'2: [-127, 128], Cr'2: [-127, 128] |
| Add bias 128 to Cb'2 and Cr'2 | |
| {Y'3, Cb'3, Cr'3} | Y'3: [0, 255], Cb'3: [1, 256], Cr'3: [1, 256] |
| Clip1Y | |
| {Y, Cb, Cr} | Y: [0, 255], Cb: [1, 255], Cr: [1, 255] |

This commit changes to this procedure:
| Representation / Operation |  Range |
| --- | --- |
| {E′R,E′G,E′B} | [0., 1.] |
| Kr-Kb like transformation with Kr = 0.25, Kb = 0.25 | |
| {Y'3, Cg'3, Co'3} | Y'3: [0., 1.], Cg'3: [-0.5, 0.5], Co'3: [-0.5, 0.5] |
| * 255 (operations from here done in `avifReformatStateUVToUNorm`) | |
| {Y'4, Cg'4, Co'4} | Y'4: [0., 255.], Cg'4: [-127.5, 127.5], Co'4: [-127.5, 127.5] |
| Add bias 128 to Cg'4 and Co'4 | |
| {Y'4, Cb'4, Cr'4} | Y'4: [0., 255.], Cb'4: [0.5, 255.5], Cr'4: [0.5, 255.5] |
| Round (rounding halfway cases up) | |
| {Y'5, Cb'5, Cr'5} | Y'5: [0, 255], Cb'5: [1, 256], Cr'5: [1, 256] |
| Clip1Y | |
| {Y, Cb, Cr} | Y: [0, 255], Cb: [1, 255], Cr: [1, 255] |

You can see that these all do the same things but in different orders.

Reply:
Thank you for the detailed reply. I know our AVIF_REFORMAT_MODE_YCGCO code doe not follow H.273's procedure. In fact, our AVIF_REFORMAT_MODE_IDENTITY code does not follow H.273's procedure, either. Ignoring rounding errors, for AVIF_REFORMAT_MODE_IDENTITY, we will still get the correct results for both full and limited ranges. For AVIF_REFORMAT_MODE_YCGCO, we will only get the correct results for full range. That's why I disallow limited range with AVIF_REFORMAT_MODE_YCGCO.